### PR TITLE
Initial implementation of plugin.

### PR DESCRIPTION
**Pull request by gwsii**

### CODE REVIEW

**Comment by gwsii** [comment]
Merging this in order to get it published to NPM.

File: lib/index.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  attachManagedPolicy() {
+    // Bail if not policies are to be applied.
+    if (!(this.serverless.service.provider.managedPolicyArns)) return
+    if (!(this.serverless.service.provider.compiledCloudFormationTemplate.Resources)) return
+
+    const { cli } = this.serverless
+    const policies = this.serverless.service.provider.managedPolicyArns
+    const resources = this.serverless.service.provider.compiledCloudFormationTemplate.Resources
+
+    cli.log('Adding managed policies...')
```
Comment (erikerikson) [inline lib/index.js]: it's a little weird to have these not corresponding to actual activity in the code.  you could assign the result of filter to a constant and then call forEach on that constant.

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    cli.log('Searching for roles...')
```
Comment (erikerikson) [inline lib/index.js]: technically, you search for roles before you add policies

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    })
+  }
+
+  /**
+   * Handler for the Serverless before:deploy:deploy event.
+   * Attaches the Managed Policy or Policies defined at `provider.managedPolicyArns`
+   * to each of the roles in the service.
+   */
+  attachManagedPolicy() {
+    // Bail if not policies are to be applied.
```
Comment (erikerikson) [inline lib/index.js]: might be better to separate validation so that it's quite clear.

File: lib/index.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    if (!(this.serverless.service.provider.managedPolicyArns)) return
+    if (!(this.serverless.service.provider.compiledCloudFormationTemplate.Resources)) return
+
```
Comment (erikerikson) [inline lib/index.js]: perhaps add validation of the config before proceeding?

File: .travis.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+language: node_js
+node_js:
```
Comment (erikerikson) [inline .travis.yml]: explicitly ignoring LTS 4.x?

File: lib/index.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+   * Given a CFT role object, apply the list of policy ARNs to the role as Managed Policies.
+   * @param {string|string[]} policies  Array of policy ARNs.
+   * @param {Object}          role      CFT Role Resource to add the policies.
+   */
+  applyPoliciesToRole(policies, role) {
+    const { cli } = this.serverless
+
+    policies.forEach((policyArn) => {
+      // Valid Policy ARN?
+      if (!policyArn.match(/^arn:aws:iam::[0-9]+:policy\/.*$/)) {
```
Comment (erikerikson) [inline lib/index.js]: validation at invocation (or even construction?) seems most appropriate

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    }
+  }
+
+  /**
+   * Given a CFT role object, apply the list of policy ARNs to the role as Managed Policies.
+   * @param {string|string[]} policies  Array of policy ARNs.
+   * @param {Object}          role      CFT Role Resource to add the policies.
+   */
+  applyPoliciesToRole(policies, role) {
+    const { cli } = this.serverless
```
Comment (erikerikson) [inline lib/index.js]: this is multiply declared.  perhaps: `this.log = this.serverless.cli.log` and then `this.log(...)` throughout, deleting these declarations?

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      if (role.ManagedPolicyArns && role.ManagedPolicyArns.indexOf(policyArn) !== -1) {
+        cli.log(`${role.RoleName} role already has policy ${policyArn} applied, skipping.`)
+        return
+      }
+
+      // Either add to list of existing policies or start a list with this one.
+      if (role.ManagedPolicyArns) {
+        cli.log(`Adding ${policyArn} to existing ManagedPolicyArns policies for ${role.RoleName}.`)
+        // eslint-disable-next-line no-param-reassign
+        role.ManagedPolicyArns = role.ManagedPolicyArns.concat(policyArn)
```
Comment (erikerikson) [inline lib/index.js]: why not just push?
also, why not do this in a bulk manner, assigning policies if `!role.ManagedPolicyArns` and filtering policies for existing policies and then using `Array.prototype.push.apply(role.ManagedPolicyArns, policies)`

File: .travis.yml (hunk 5 above)
Comment (gwsii) [reply .travis.yml]: Not intending to. Don't think Travis is quite set up correctly yet, either. Maybe something else I need to do.

File: lib/index.js [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   * Create an instance of our ServerlessPlugin.
+   * @param   {Object} serverless Host Serverless instance into which this
+   *                              plugin is loading.
+   * @returns {Object} Instance of this plugin for use by Serverless.
+   */
+  constructor(serverless) {
+    this.serverless = serverless
+    this.hooks = {
+      'before:deploy:deploy': this.attachManagedPolicy.bind(this),
```
[the remark this answers is not in the document]
Comment (gwsii) [reply lib/index.js]: Not sure. I'll have to test it. Seems like a bigger issue for timing between plugins and agree this needs to be explored.

File: lib/index.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+   */
+  // eslint-disable-next-line class-methods-use-this
+  policiesArray(policies) {
+    // Must be a string for a single policy or an array for multiple.
+    if (typeof policies === 'string') {
+      return [policies]
+    } else if (policies instanceof Array) {
+      return policies
+    } else {
+      throw new Error('managedPolicyArns must be an array')
```
[the remark this answers is not in the document]
Comment (gwsii) [reply lib/index.js]: Improved the message.

[the remark this answers is not in the document]
Comment (gwsii) [reply lib/index.js]: Pushed all validation to the front of the invocation.

File: lib/index.js (hunk 7 above)
Comment (gwsii) [reply lib/index.js]: Did that.

File: lib/index.js (hunk 6 above)
Comment (gwsii) [reply lib/index.js]: pushed to invocation

File: lib/index.js (hunk 8 above)
Comment (gwsii) [reply lib/index.js]: Yeah, that method was a bit much. Too chatty and a bit smelly. Got it down to four lines.

File: lib/index.js (hunk 3 above)
Comment (gwsii) [reply lib/index.js]: Still include these two checks to bypass entire logic, but then all other validation now occurs here.

File: lib/index.js (hunk 4 above)
Comment (gwsii) [reply lib/index.js]: Yep

File: lib/index.js (hunk 1 above)
Comment (gwsii) [reply lib/index.js]: Reduced and simplified the output.

File: lib/index.js (hunk 2 above)
Comment (gwsii) [reply lib/index.js]: Fixed.

File: .travis.yml (hunk 5 above)
Comment (gwsii) [reply .travis.yml]: This is sorted. Added 4 LTS and latest node and enabled Travis CI for repo.

**Pull request merged**